Prevents Nether Portals from being created if they would touch another

megagamer410 opened the pull request:
nether portal. Consistent with (half) of default minecraft functionality, and prevents a rendering error where having 2 nether portals facing the same direction touching each other would cause the portals to become invisible.

Signed-off-by: megagamer410 megagamer410@gmail.com

DDoS commented on src/main/java/org/spout/vanilla/material/block/portal/NetherPortal.java:
```diff
@@ -99,10 +99,23 @@ public boolean createPortal(Block bottomBlock) {
 		}
-		// Validated, time to create the portal
+		// Validate that there are no portals touching the proposed one.
+		// NOTE: this matches the default minecraft functionality only on portals facing North-South
```
Yeah, not sure about the half compatibility thing. It would be better to do it now entirely, instead of doing it partially, forgetting about it, and then finally fixing it later...

megagamer410 replied:
The problem with that is that the default minecraft implementation isn't particularly logical. If the portals face North / South then it doesn't create the new one, but if it's East / West then it creates the new one, unless it's being created between 2 existent portals, at which case it destroys the west most one and fails to build the proposed one. I opted to only commit the North / South implementation because it makes the most sense, but I'll start working on the East / West compatibility.

Raphfrk replied:
I don't think we need an exact match if it makes no sense.  @zidane can comment.

Does this also include creating portals in other worlds?

megagamer410 replied:
Portal creation is identical in the Overworld and the Nether but fails in the End, so I updated the commit to reflect that.

windy1 replied:
Are nether portals supposed to work in the end or is a bug in Minecraft?

megagamer410 replied:
Nether portals are not supposed to work in the end.

windy1 commented on src/main/java/org/spout/vanilla/material/block/portal/NetherPortal.java:
```diff
@@ -56,6 +57,10 @@ public MoveReaction getMoveReaction(Block block) {
 	 */
 	public boolean createPortal(Block bottomBlock) {
+		//  Check if the portal is being created in the End.
+		if (bottomBlock.getBiomeType().getClass().equals(EndStoneBiome.class)) {
```
I think

``` java
if (world.getGenerator() instanceof TheEndGenerator) {
    // do stuff
```

would be more logical here.

windy1 commented:
And I presume you've tested this as well? :p

megagamer410 commented on src/main/java/org/spout/vanilla/material/block/portal/NetherPortal.java:
```diff
@@ -56,6 +57,10 @@ public MoveReaction getMoveReaction(Block block) {
+		//  Check if the portal is being created in the End.
+		if (bottomBlock.getWorld().getGenerator() instanceof TheEndGenerator) {
+			return false;
+		}
```
This prevents Nether Portals from being created in the End, but leaves the fire. (Same as Minecraft)

megagamer410 commented on src/main/java/org/spout/vanilla/material/block/portal/NetherPortal.java:
```diff
@@ -99,6 +104,21 @@ public boolean createPortal(Block bottomBlock) {
+				}
+			}
+			corner1 = corner1.translate(BlockFace.TOP);
+			corner2 = corner2.translate(BlockFace.TOP);
```
The 3 iteration for loop, combined with these 2 lines fixes a bug in Minecraft that allows players to create 2 nether portals that are touching, causing some of both portals to become invisible. (Minecraft Bug Fix)

megagamer410 commented on src/main/java/org/spout/vanilla/material/item/tool/FlintAndSteel.java:
```diff
@@ -68,6 +63,11 @@ public void onInteract(Entity entity, Block block, Action type, BlockFace clicke
+				// Handle the creation of portals
+				if (VanillaMaterials.PORTAL.createPortal(target.translate(BlockFace.BOTTOM))) {
+					return;
+				}
```
Portal creation has to occur after the fire is placed so that the fire can be extinguished by the portal creation sequence when the obsidian frame is proper but the portal can't be generated due to a conflict with an already existing portal.

megagamer410 commented:
Updated the commit to fix a Minecraft bug and to include W1ndwaker's suggestion.

DDoS commented on src/main/java/org/spout/vanilla/material/block/portal/NetherPortal.java:
```diff
@@ -99,6 +104,21 @@ public boolean createPortal(Block bottomBlock) {
+			// Check if there is another nether portal touching the proposed one
+			for (BlockFace side : BlockFaces.NESW) {
+				// If so, don't build the current one, and extinguish the fire.
+				if (corner1.translate(side).getMaterial().equals(VanillaMaterials.PORTAL) || corner2.translate(side).getMaterial().equals(VanillaMaterials.PORTAL)) {
```
switch:
corner1.translate(side).getMaterial().equals(...
with:
corner1.translate(side).isMaterial(...
 and I'll merge this (same thing with corner2).

megagamer410 committed:
Prevents Nether Portals from being created if they would touch another
nether portal. Also resolves bug in Minecraft that allows 2 nether portals
to touch if the y-difference between them is 2. Prevents Nether Portals from
lighting in the End.

Signed-off-by: megagamer410 <megagamer410@gmail.com>
